From 4b45abcdb107c8c5d58fd63a2616c54c800305b3 Mon Sep 17 00:00:00 2001
From: Akhilesh Samineni <akhilesh.samineni@broadcom.com>
Date: Mon, 5 Apr 2021 13:21:40 -0700
Subject: [PATCH] Link local scope was not set while binding socket with local
 address causing socket errors for bgp ipv6 link local neighbors.

---
 bgpd/bgp_network.c | 5 +++++
 bgpd/bgp_zebra.c   | 3 +++
 2 files changed, 8 insertions(+)

diff --git a/bgpd/bgp_network.c b/bgpd/bgp_network.c
index 6a5c2c4b3..d7047a154 100644
--- a/bgpd/bgp_network.c
+++ b/bgpd/bgp_network.c
@@ -556,6 +556,11 @@ static int bgp_update_address(struct interface *ifp, const union sockunion *dst,
 		return 1;
 
 	prefix2sockunion(sel, addr);
+
+        if (IN6_IS_ADDR_LINKLOCAL(&addr->sin6.sin6_addr)) {
+                addr->sin6.sin6_scope_id = ifp->ifindex;
+        }
+
 	return 0;
 }
 
diff --git a/bgpd/bgp_zebra.c b/bgpd/bgp_zebra.c
index e42d6ee26..8e9fa8903 100644
--- a/bgpd/bgp_zebra.c
+++ b/bgpd/bgp_zebra.c
@@ -814,6 +814,9 @@ bool bgp_zebra_nexthop_set(union sockunion *local, union sockunion *remote,
 								? peer->conf_if
 								: peer->ifname,
 							peer->bgp->vrf_id);
+			else if (peer->update_if)
+				ifp = if_lookup_by_name(peer->update_if,
+						peer->bgp->vrf_id);
 		} else if (peer->update_if)
 			ifp = if_lookup_by_name(peer->update_if,
 						peer->bgp->vrf_id);
-- 
2.18.0

